kernel/src/transaction.rs

```diff
@@ -1,120 +1,249 @@
+        let commit_path = &self
+            .read_snapshot
+            .table_root
+            .join("_delta_log/")?
+            .join(&commit_file_name)?;
+
+        // step three: commit the actions as a json file in the log
+        let json_handler = engine.get_json_handler();
+        match json_handler.write_json_file(commit_path, Box::new(actions), false) {
+            Ok(()) => Ok(CommitResult::Committed(commit_version)),
+            Err(crate::error::Error::ObjectStore(object_store::Error::AlreadyExists {
+                ..
+            })) => Ok(CommitResult::Conflict(self, commit_version)),
+            Err(e) => Err(e),
+        }
     }
 
-    /// Scan the Delta Log to obtain the latest transaction for all applications
-    pub fn application_transactions(&self, engine: &dyn Engine) -> DeltaResult<TransactionMap> {
-        self.scan_application_transactions(engine, None)
+    /// Add commit info to the transaction. This is commit-wide metadata that is written as the
+    /// first action in the commit. Note it is required in order to commit. If the engine does not
+    /// require any commit info, pass an empty `EngineData`.
+    pub fn commit_info(&mut self, commit_info: Box<dyn EngineData>, schema: Schema) {
+        self.commit_info = Some(EngineCommitInfo {
+            data: commit_info.into(),
+            schema: schema.into(),
+        });
     }
 }
 
-#[cfg(all(test, feature = "default-engine"))]
-mod tests {
-    use std::path::PathBuf;
+/// Result after committing a transaction. If 'committed', the version is the new version written
+/// to the log. If 'conflict', the transaction is returned so the caller can resolve the conflict
+/// (along with the version which conflicted).
+pub enum CommitResult {
+    /// The transaction was successfully committed at the version.
+    Committed(crate::Version),
+    /// The transaction conflicted with an existing version (at the version given).
+    Conflict(Transaction, crate::Version),
+}
+
+// given the engine's commit info (data and schema as [EngineCommitInfo]) we want to create both:
+// (1) the commitInfo action to commit (and append more actions to) and
+// (2) the schema for the actions to commit (this is determined by the engine's commit info schema)
+#[allow(clippy::type_complexity)]
+fn generate_commit_info(
+    engine: &'_ dyn Engine,
+    engine_commit_info: Option<EngineCommitInfo>,
+) -> DeltaResult<(
+    Box<dyn Iterator<Item = Box<dyn EngineData>> + Send + '_>,
+    SchemaRef,
+)> {
+    use crate::actions::{
+        ADD_NAME, COMMIT_INFO_NAME, METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME, TRANSACTION_NAME,
+    };
+
+    // TODO: enforce single row commit info
+    // TODO: for now we always require commit info
+    let action_schema = Arc::new(engine_commit_info.as_ref().map_or(
+        get_log_schema().clone(),
+        |commit_info| {
+            let mut action_fields = get_log_schema().fields().collect::<Vec<_>>();
```

Comment:
This is a bit much, but if `fields()`'s signature was changed to return a `DoubleEndedIterator`, you can also avoid materializing this, use `next_back`, and remove the `into_iter` down below. Tho I imagine `action_fields` is pretty small anyway.

```rust
                    let mut action_fields = get_log_schema().fields();
                    let commit_info_field = action_fields
                        .next_back()
                        .expect("last field is commit_info in action schema");
                    //...
                    let action_fields = action_fields.cloned().chain(std::iter::once(
                        crate::schema::StructField::new(COMMIT_INFO_NAME, commit_info_schema, true),
                    ));
                    StructType::new(action_fields.collect())
```

Reply:
haha nice I like this and will keep open to consider in the future but probably punting on it in this PR?